src/Http/Middleware/CspMiddleWare.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.6.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Middleware;
+
+use Cake\Core\Configure;
+use Cake\Http\Response;
+use Cake\Http\ServerRequest;
+use InvalidArgumentException;
+use ParagonIE\CSPBuilder\CSPBuilder;
+
+/**
+ * Content Security Policy Middleware
+ */
+class CspMiddleware
+{
+
+    /**
+     * CSP Builder
+     *
+     * @var \ParagonIE\CSPBuilder\
+     */
+    protected $csp;
+
+    /**
+     * Constructor
+     *
+     * @param null|\ParagonIE\CSPBuilder\
+     */
+    public function __construct($csp = null)
+    {
+        if (!is_array($csp)) {
+            $cspConfig = (array)Configure::read('App.CSP');
```

Comment:
Why would this be an application config value? Do you see the CSP header varying by environment?

Reply:
In general too let's not make new classes directly read from `Configure::read()` if possible. Even if someone is saving the setting in `Configure` they can easily do `new CspMiddleware(Configure::read('App.CSP'))`.

Reply:
It's not creating a new class based on it, it's just passing the configuration to the new instance.

Reply:
There's no problem using `Configure` to store configure. But there's no need to use `Configure::read()` inside the class constructor where you can easily read and pass the config array as argument.

For datasouces also we do the [same](https://github.com/cakephp/app/blob/master/config/bootstrap.php#L146). Read the config array and pass it to `ConnectionManager::setConfig()`.

Reply:
Ah OK, I get your point now. But the problem is then, even if we provide defaults, the user would still have to be aware of them and manually pass them. I was looking for a convenient way to make CSP work out of the box with some good default settings.

Reply:
We can make it work out of the box by updating the application skeleton as needed.